use libzedmd 0.9.1 for ZeDMD 5.1.x

mkalkbrenner opened the pull request.

freezy commented:
I'm not sure why you've removed the RenderColoredGrayXXX APIs? Existing colorizations made for 16M colors won't render the same as for 65k colors..

mkalkbrenner commented:
I removed anything but RGB565 zone streaming in libzedmd. ZeDMD Wifi never supported anything else. Now with ZeDMD 5.0.0 I removed all the different modes that were only supported over USB. 
In DMDext 2.3.0 we already used RGB565 zone streaming for RGB888 content. So that isn't new.

freezy commented:
I worry  that colorizations using 64 colors out of 16 million colors will now get 64 colors out of 65K colors. This is breaking BC, in my opinion, since colors will be different than before. 

And it concerns the vast majority of ZeDMD users, and the vast majority of colorizations. I'm not talking about the new RGB888 content.

mkalkbrenner commented:
Since ZeDMD WiFi has been released nobody complained about the colors. And I can't see a difference in the colors.
Zone Streaming gives us many advantages in the firmware. As mentioned before, we never used any other mode with WiFi and therefore iOS or Android. And I don't want to continue maintaining different branches.

BTW I'm currently experimenting with changing the SD variations into fixed size devices.

freezy commented:
Well, "nobody complained" isn't very re-assuring, since looking at the code, there isn't a RGB565 fallback for VNI colorizations, meaning they won't work at all, or will fall back to RGB24, which will lag...

I'll need more testing for this, and comparison shots. I know it's tempting to just move forward and drop support for legacy tech, but this is a software that thousands of people are using, and breaking BC is a big deal.

I don't know what Zone Streaming is, if it's relevant I'd like to hear more about it.

mkalkbrenner commented:
For sure, VNI works. I always test with Batman (for dynamic 128x16 to 128x32 changes) and Baywatch (192x64) VNI colorizations.

And yes, within DMDext it uses RGB24. Within libzedmd, RGB888 gets converted in RGB565.

mkalkbrenner commented:
Again, the new implementation basically uses the WiFi implementation we had in DMDext 2.3.0 for USB as well.
And in DMDext 2.3.0, we already sent RGB24 as RGB565, converted within libzedmd.

So there's no change for at all for WiFi users. And for USB users we previously had the experimental enforceStreming method. We already used that on other operating systems. Now it is time to drop all the legacy code and to focus on maintainablility for the future.

BTW I don't think that we have thousands of ZeDMD users ;-)

freezy commented:
Sorry for having to bear with me, but many things are still not clear:

- You keep bringing up the WiFi implementation. I don't understand the relevance of the WiFi implementation. For me, WiFi is a hardware component on the board, so given it's there, you added support for it. But I doubt anyone in their right mind would actually use it in a cabinet when they can simply plug it in via USB, which is more stable by design and not dependent on any external factors, and you need to plug it in somewhere anyway to power it. Am I missing something here?
- So the "there is no change for WiFi users" argument is kind of weak, IMO. How many do you think there are, compared to USB users?
- I don't know what "enforceStreaming" and the Zone thing above mean. I'm not closely following libzedmd development, so if you're introducing new features, you'll need to explain or document them somewhere so I can read about them.

I've been maintaining ~10 different frame formats for multiple displays for nearly a decade now. We have VNI (and Serum v1) that still produce colored-gray-x frames, so I don't consider this format legacy. Making the conversion go `ColoredGray6` -> `RBG24` -> `RGB565` is technically inferior to sending as `coloredGray4` directly. You're doing two unnecessary conversions, and while I don't have any metrics on whether this has an actual impact, I find it an unappealing approach.

That being said, if it has no impact performance-wise, and if the result is identical, I'm willing to give it a shot, but I'll need more proof than "nobody complained". By experience, testing internally rarely aligns with what people do with it once officially released.

freezy commented:
My thoughts about the forced RGB24 -> RGB565 conversion: I didn't like it in v2.3.0 neither, because it reduces the color fidelity. Since this was only relevant for still images, I accepted it because still images with gradients that where it's actually visible are probably pretty rare. So, while it as breaking BC, it applied to very few cases.

Now, these changes apply to literally every image sent to ZeDMD. So I really want to make sure the results are the same and indistinguishable from those before, because I can 100% assure you that people will complain about it if that's not the case.

mkalkbrenner commented:
The USB to serial converter used on the ESP32 is simply not fast enough to transfer RGB24 full frames.
Even with 64 color modes, some frames were dropped in fast animations or rendered with delay.

WiFi is much faster and therefore the preferred way to transport the data. Especially for 256x64 builds.

In opposite to other and likely more expensive devices, we need to be smart instead of just sending raw frames or planes.
Therefore, data gets compressed. But there's not enough memory on an ESP32 to uncompress entire frames.
That's why I implemented RGB565 zone streaming.
libzedmd divides a complete frame into 128 zones and treats each zone individually and only transmitts zones that changed compared to the precious frame. The data of such a zone is encoded in RGB565.

In detail, there're more small tricks used to increase the performance. But the important part is that ZeDMD HD, ZeDMD WiFi and ZeDMD HD WiFi always used that advanced impelementation for any mode.

Only ZeDMD supported "native" grayscale based modes because that were the modes we initially supported. But that device was very bad on displaying RGB24. With Serum v2 and the good reusults when using RGB565, it seemed obvious to drop all these legacy modes and to use RGB565 zone streaming for everything, just like ZeDMD HD, ZeDMD WiFi and ZeDMD HD WiFi always did.

mkalkbrenner commented:
Using this PR, we mainly move tasks like scaling and conversion to RGB24 to DMDext instead of doing it in libzedmd.
We did the same with libdmdutil to have the same code doing such things for different devices like Pixelcade, ZeDMD, etc.
DMDext already has the capabilities already, in libdmdutil we added them.

zesinger commented:
For the difference between RGB888 and RGB565, you could perform the test under ColorizingDMD as I coded my own color picker window and there you can set either a RGB888 or a RGB565 color and see what it looks like in the other resolution in the top right square (divided into 2).
https://i.servimg.com/u/f88/17/72/22/03/screen68.png
Sincerely, in very rare occasions, you can **discern** a small difference. It won't change anything for an image, the gradients must be have very slight changes for RGB565 to be unable to reproduce it. And that is on a modern PC monitor, now if you transpose that on a dot matrix display, I'm 99% sure you can not detect any difference.

PastorL69 commented:
Been following this for a little bit but as an author myself I can of course see very slight differences. Although what matters most for me is the performance. I'd rather have a fluid colorization running than 1:1 accurate colors that come with frequent stutters. Not to forget, ZeDMD runs on 7 or 8bit, so you lose color accuracy no matter what. 

From what I understand, this being exclusive to ZeDMD seems like a great move. Adopting RGB565 as the new standard feels like the right step forward if we want everything to run as it should.

mkalkbrenner commented:
libzedmd 0.8.0 and ZeDMD 5.0.0 have been released. Both will be incompatible to dmdext until this PR gets merged.
Meanwhile you can use ZeDMD 5.0.0 with these DMDext DLLs:
https://github.com/PPUC/dmd-extensions/releases/tag/v2.3.1-zedmd.beta.1

mkalkbrenner commented:
@freezy libzedmd 0.9.0 and ZeDMD 5.1.0 have been officially released.
libdmdutil and VPX Standalone are already using it.

I think that the API is stable now for a longer time. And since we have the version numbers now, we can also create backward compatibility features in the future.

It would be great to get that into DMDext before too many people update their ZeDMD or PINDMD and start complaining that it isn't working.

Meanwhile, people could use these DLLs:
https://github.com/PPUC/dmd-extensions/releases/tag/v2.3.1-zedmd.beta.3

Gerhard-BL commented:
Just installed latest ZeDMD 5.1.x 
It's working really good, even with VPX-STANDALONE (Android)
But it would be great, if required Libs could be implemented into next Release of extDMD for Windows VPin, so there is no need to install "special Versions" for ZeDMD.
Also it would be a lot easier, solve others Issues, because talking about the same Base.

MikedaSpike commented:
As a regular end user, I do not comprehend the difference between RGB565 and RGB888. 
I use the latest firmware (5.1.1) on my ZeDMD HD with this PR, and the media I checked appears normal. I can play 4-color, 16-color, 64-color, and even full RGB media on it without noticing any issues. It can even display very fast frames and works flawlessly!

mkalkbrenner commented:
https://github.com/PPUC/dmd-extensions/releases/tag/v2.3.1-zedmd.beta.4

G-Diddy01 commented:
I was having issue with my PinDMDv3 with Pinup Popper.  Popper would give exception error on every table exit or would pause 30 seconds before buttons would respond again for every table. When I applied ZeDMD 5.1.2 it fixed the issues and it runs well on my PinDMDv3 with my dmdext 2.3.1-beta.4 build.  Im requesting this fix makes it into the next build!   Im so happy this has a fix after what seems like 5 months battling this issue. Also up to date with VPinMame.  Thanks!

PastorL69 commented:
Merge Merge Merge 😄

freezy commented:
![image](https://github.com/user-attachments/assets/b9fa9f98-077c-4907-8d56-ac62a804313f)

Is it me or did this PR seriously not compile?
